src/rest-server/src/controllers/v2/azureAD.js

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+// module dependencies
+
+const authnConfig = require('../../config/authn');
+const querystring = require('querystring');
+const axios = require('axios');
+const createError = require('../../util/error');
+const jwt = require('jsonwebtoken');
+
+const requestAuthCode = async (req, res, next) => {
+  const tenantId = authnConfig.OIDCConfig.tenantID;
+  const clientId = authnConfig.OIDCConfig.clientID;
+  const responseType = 'code';
+  const redirectUri = authnConfig.OIDCConfig.redirectUrl;
+  const responseMode = 'form_post';
+  const scope = 'openid offline_access https://graph.microsoft.com/user.read';
+  const state = 'openpai';
+  return res.redirect('https://login.microsoftonline.com/' + tenantId + '/oauth2/v2.0/authorize?'+ querystring.stringify({
```

Comment:
I think we'd better use OIDC's well known config to get the authroize and token url.
Some Azure region may not use `https://login.microsoftonline.com/` as the login host.
e.g. mooncake: https://docs.azure.cn/zh-cn/active-directory/develop/active-directory-v2-protocols

Reply:
Graph API endpoint is also different. https://docs.azure.cn/zh-cn/active-directory/develop/active-directory-graph-api-quickstart#how-to-construct-a-graph-api-url

Reply:
I think we should create a known issue if we only supports the default endpoint.

Reply:
National Clouds endpoint list https://docs.microsoft.com/en-us/azure/active-directory/develop/authentication-national-cloud

Reply:
Done